### PR TITLE
chunked prefill

### PR DESCRIPTION
**Pull request by mailvijayasingh**
# Description

Start with a short description of what the PR does and how this is a change from
the past.
The current CL is WIP and supports Chunked Prefill for unquantized KV Cache. 

attentions.py file has mainly 2 changes-
i) Fetching previous cache is there are processed chunks
ii) Generating attention masks for a particular chunk based on article - https://medium.com/byte-sized-ai/llm-inference-optimizations-2-chunked-prefill-764407b3a67a

Minor chnages - 
wiring previous chunk information to decoder, model, attentions layers. 

The rest of the description includes relevant details and context, examples:
- why is this change being made,
- the problem being solved and any relevant context,
- why this is a good solution,
- some information about the specific implementation,
- shortcomings of the solution and possible future improvements.

If the change fixes a bug or a Github issue, please include a link, e.g.,:
FIXES: b/123456
FIXES: #123456

# Tests

Please describe how you tested this change, and include any instructions and/or
commands to reproduce.

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed.

### CODE REVIEW

File: MaxText/maxengine.py [hunk 1]
```diff
@@ -329,10 +333,13 @@ def prefill(
...
-
-    zero_to_n = jnp.arange(0, padded_tokens.shape[0])
-    ones_to_keep = zero_to_n < true_length
+    if positions is None:
+      positions = jnp.expand_dims(jnp.arange(0, input_tokens.shape[1]), 0)
+    if not self.config.use_chunked_prefill:
+      zero_to_n = jnp.arange(0, padded_tokens.shape[0])
+    else:
+      zero_to_n = jnp.arange(0, complete_prompt_padded_length.shape[0])
+    ones_to_keep = zero_to_n < complete_prompt_true_length
```
[the remark this answers is not in the document]
Comment (mailvijayasingh) [reply MaxText/maxengine.py:342]: complete_prompt_true_length - is the complete prompt (in true length, without padding), we need that to set decoder segment ids which is used in decode step. 
 
 true_length is the true length of the chunked prompt 

yes, correct, true_length == padded_length for all chunks except last chunk, to avoid if statement to check last chunk or not, I preferred to pass true_length. Let me know if you have concerns?

File: MaxText/layers/attentions.py [hunk 2]
```diff
@@ -786,11 +817,30 @@ def kv_cache_prefill(
     cached_prefill_key_vars[0].value = key_shaped_for_cache
     cached_prefill_value_vars[0].value = value_shaped_for_cache
 
+    if self.config.use_chunked_prefill:
```
[the remark this answers is not in the document]
Comment (mailvijayasingh) [reply MaxText/layers/attentions.py:820]: sure, will send another version with updated comments.

File: MaxText/configs/base.yml [hunk 3]
```diff
@@ -564,3 +564,7 @@ sa_use_fused_bwd_kernel: False
 sa_q_layout: "HEAD_DIM_MINOR"
 sa_k_layout: "HEAD_DIM_MINOR"
 sa_v_layout: "HEAD_DIM_MINOR"
+
+# Chunked Prefill Parameters
+chunk_size: 256
```
Comment (vipannalla) [inline MaxText/configs/base.yml:569]: Please rename this to `prefill_chunk_size`. `Chunk_size` is too generic, paged attention also has a chunk size :).

File: MaxText/maxengine.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (vipannalla) [reply MaxText/maxengine.py:342]: As talked offline, please update method docbook with this info and an example.

File: MaxText/layers/attentions.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (mailvijayasingh) [reply MaxText/layers/attentions.py]: done

File: MaxText/configs/base.yml (hunk 3 above)
Comment (mailvijayasingh) [reply MaxText/configs/base.yml]: done

**Review by vipannalla — APPROVED** [review]
looks good

**Review by khatwanimohit — APPROVED** [review]
